### PR TITLE
BFE support

### PR DESCRIPTION
**Pull request by arj03**
This PR makes it possible to use envelope-js with BFE. Most of the changes are minor code formatting. There are only 2 real changes here:

1) Use bfe.json instead of tfk.json

2) Handle null values for messages. This change can seem a bit counter-intuitive coming from TFK. In BFE there is only 1 nil type. This means that a null message id will be encoded as `[6 2]` no matter what feed format are. There only place where this is important in the tribes & envelop-js libaries and that is in derive secret, so we just need to distinguish the feed format there.

Note this adds ssb-binary-field-encodings-spec from github to package.json. I'll do a proper version on npm once we are ready for that.

### CODE REVIEW

File: box.js [hunk 1]
```diff
@@ -13,8 +13,8 @@ module.exports = function box (plain_text, feed_id, prev_msg_id, msg_key, recp_k
   const derive = DeriveSecret(feed_id, prev_msg_id)
 
   const read_key = derive(msg_key, [LABELS.read_key])
-    const header_key = derive(read_key, [LABELS.header_key])
-    const body_key   = derive(read_key, [LABELS.body_key])
```
Comment (mixmix) [inline box.js:17]: These indents were reminders to the developer about the heirarchy of these keys.
I'd like to leave it like this (or similar) to make the code easier to read.

I would also be fine with putting a comment in if that's better:

```js
// read_key
//   ├──> header_key
//   └──> body_key
```

File: package.json [hunk 2]
```diff
@@ -18,7 +18,9 @@
     "buffer-xor": "^2.0.2",
     "envelope-spec": "^1.0.0",
     "futoin-hkdf": "^1.3.2",
-    "sodium-native": "^3.2.0"
+    "sodium-native": "^3.2.0",
+    "ssb-bfe": "^0.5.0",
```
Comment (mixmix) [inline package.json:22]: looks like this is a devDep at the moment?
Is that right?

File: package.json [hunk 3]
```diff
@@ -18,7 +18,9 @@
...
+    "ssb-binary-field-encodings-spec": "github:ssb-ngi-pointer/ssb-binary-field-encodings-spec"
```
Comment (mixmix) [inline package.json:23]: TODO, reference a version `^1.0.0` of this before merging

File: test/bfe-nil-msg.test.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  t.equal(classicNull.slice(0,2).toString('hex'), '0100', 'classic msg')
+  t.ok(classicNull.slice(2).equals(Buffer.alloc(32)), 'ends with 32 zeros')
+  t.equal(classicNull.length, 32 + 2, 'length ok')
+
+  const msg = '%HZVnEzm0NgoSVfG0Hx4gMFbMMHhFvhJsG2zK/pijYII=.sha256'
+  const classicMsg = bfe.convertNilForEnvelopeSpec(bfe.encode(msg), bfe.encode(classicFeed))
+  t.equal(classicMsg.slice(0,2).toString('hex'), '0100', 'classic msg')
+  t.notOk(classicMsg.slice(2).equals(Buffer.alloc(32)), 'is not null')
+  t.equal(bfe.decode(classicMsg), msg, 'decodes back to the same')
+  t.end()
```
Comment (mixmix) [inline test/bfe-nil-msg.test.js:23]: I don't understand why this test is in this repo?
Also why is it that you would ever want an empty data part for these types?

 # :fire: (i.e. danger)

I feel like you should not be allowed to make zero-data / zero-key feed/msg?
Seems like a good way to shoot yourself in the foot.
BUUUT, I also don't have the full context loaded into my head

**Review by mixmix — COMMENTED** [review]
## Seems good to me!
Some minor changes requested/ questions asked.

Like that you have a BFE module with helpers
For convenience I wonder if it might be nice to include that spec inside the ssb-bfe, then allow you to access the spec like `const { bfeTypes } = require('ssb-bfe'). Then the helper and spec will always be aligned

## Missing
We're going to need associated changes to https://github.com/ssbc/envelope-spec/ which references TFK

File: box.js (hunk 1 above)
Comment (staltz) [reply box.js:17]: Yeah a comment would be much more clear. I didn't see it from the indentation sand it just looked like incorrect indentation

File: package.json (hunk 2 above)
Comment (arj03) [reply package.json:22]: Right, thanks!

File: test/bfe-nil-msg.test.js (hunk 4 above)
Comment (arj03) [reply test/bfe-nil-msg.test.js:23]: This is the null case I alluded to. Before (in ssb-bfe) when encoding a message, say `{ subfeed: x.ed25519, metafeed: y.bbfeed-v1, previous: null }` you have to use the correct encoder as it's not really possible to detect from the content what previous is here. This is error prone and not super general. Instead now you just call encode and decode and it works. We didn't add this kind of conversion code to bfe as it is not really a concern of that library. Bendy butt works perfectly fine with the same nil message encoding for classic and bendy butt messages. I came from reading the TFK spec first so I thought this wouldn't work, but we really only need to handle it here, in ssb-tribes there are no changes related to this.

Comment (mixmix) [reply test/bfe-nil-msg.test.js:23]: Ah yeah, I remember. This is about the specific very unlikely edge case that prev_msg_id is null
(unlikely because starting with an encrypted message is fairly rare, but not impossible)
Yes?

Comment (mixmix) [reply test/bfe-nil-msg.test.js:23]: I don't understand the Meta feed context enough to fully grok what you've said.

Comment (arj03) [reply test/bfe-nil-msg.test.js:23]: yes

**Comment by arj03** [comment]
Great review @mixmix. I pushed up a change for the issues raised. Adding the helper to bfe was a really good idea. Now ssb-bfe uses that bfe.json internally as well.

I have created a PR for the envelope spec here: https://github.com/ssbc/envelope-spec/pull/21

**Comment by mixmix** [comment]
awesome. This looks ready to merge and publish to me <3
thanks @arj03
